pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarPartitionLevelConsumer.java

```diff
@@ -45,93 +40,66 @@
+              partition), e);
     }
-    LOGGER.info("Created Pulsar reader with topic: {}, partition: {}, initial message id: {}",
-        _config.getPulsarTopicName(), partitionId, _config.getInitialMessageId());
+    LOGGER.info("Created Pulsar reader for topic: {}, partition: {}", topicName, partition);
   }
 
-  /**
-   * Fetch records from the Pulsar stream between the start and end StreamPartitionMsgOffset
-   * Used {@link org.apache.pulsar.client.api.Reader} to read the messaged from pulsar partitioned topic
-   * The reader seeks to the startMsgOffset and starts reading records in a loop until endMsgOffset or timeout is
-   * reached.
-   */
   @Override
-  public PulsarMessageBatch fetchMessages(StreamPartitionMsgOffset startMsgOffset, int timeoutMs) {
-    MessageIdStreamOffset startOffset = (MessageIdStreamOffset) startMsgOffset;
+  public synchronized PulsarMessageBatch fetchMessages(StreamPartitionMsgOffset startOffset, int timeoutMs) {
+    MessageId startMessageId = ((MessageIdStreamOffset) startOffset).getMessageId();
+    long endTimeMs = System.currentTimeMillis() + timeoutMs;
     List<BytesStreamMessage> messages = new ArrayList<>();
-    Future<PulsarMessageBatch> pulsarResultFuture = _executorService.submit(() -> fetchMessages(startOffset, messages));
-    try {
-      return pulsarResultFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
-    } catch (TimeoutException e) {
-      // The fetchMessages has thrown an exception. Most common cause is the timeout.
-      // We return the records fetched till now along with the next start offset.
-      pulsarResultFuture.cancel(true);
-    } catch (Exception e) {
-      LOGGER.warn("Error while fetching records from Pulsar", e);
+
+    // Seek to the start message id if necessary
+    // NOTE: Use Objects.equals() to check reference first for performance.
+    if (!Objects.equals(startMessageId, _nextMessageId)) {
+      try {
+        _reader.seek(startMessageId);
+      } catch (PulsarClientException e) {
+        throw new RuntimeException("Caught exception while seeking to message id: " + startMessageId, e);
+      }
     }
-    return buildPulsarMessageBatch(startOffset, messages);
-  }
 
-  private PulsarMessageBatch fetchMessages(MessageIdStreamOffset startOffset, List<BytesStreamMessage> messages) {
+    // Read messages until all available messages are read, or we run out of time
     try {
-      MessageId startMessageId = startOffset.getMessageId();
-      _reader.seek(startMessageId);
-      while (_reader.hasMessageAvailable()) {
-        Message<byte[]> message = _reader.readNext();
-        messages.add(PulsarUtils.buildPulsarStreamMessage(message, _config));
-        if (Thread.interrupted()) {
-          break;
-        }
+      while (_reader.hasMessageAvailable() && System.currentTimeMillis() < endTimeMs) {
+        messages.add(PulsarUtils.buildPulsarStreamMessage(_reader.readNext(), _config));
```

Comment:
Assuming that readNext will block.

Reply:
`readNext()` shouldn't block. We call `readNext()` only when there `hasMessageAvailable()`

Reply:
> `readNext()` shouldn't block. We call `readNext()` only when there `hasMessageAvailable()`
Makes sense if hasMessageAvailable is used. I read this and thought it blocks
https://github.com/apache/pulsar/blob/master/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Reader.java#L49